pkg/proxy/conntrack/cleanup.go

```diff
@@ -20,95 +20,109 @@ limitations under the License.
-		if svcInfo, ok := svcPortMap[svcPortName]; ok {
-			klog.V(4).InfoS("Newly-active UDP service may have stale conntrack entries", "servicePortName", svcPortName)
-			conntrackCleanupServiceIPs.Insert(svcInfo.ClusterIP().String())
-			for _, extIP := range svcInfo.ExternalIPs() {
-				conntrackCleanupServiceIPs.Insert(extIP.String())
-			}
-			for _, lbIP := range svcInfo.LoadBalancerVIPs() {
-				conntrackCleanupServiceIPs.Insert(lbIP.String())
-			}
-			nodePort := svcInfo.NodePort()
-			if svcInfo.Protocol() == v1.ProtocolUDP && nodePort != 0 {
-				conntrackCleanupServiceNodePorts.Insert(nodePort)
-			}
-		}
-	}
+	start := time.Now()
+	klog.V(4).InfoS("Started to reconcile conntrack entries", "ipFamily", ipFamily)
 
-	klog.V(4).InfoS("Deleting conntrack stale entries for services", "IPs", conntrackCleanupServiceIPs.UnsortedList())
-	for _, svcIP := range conntrackCleanupServiceIPs.UnsortedList() {
-		filters = append(filters, filterForIP(svcIP, v1.ProtocolUDP))
-	}
-	klog.V(4).InfoS("Deleting conntrack stale entries for services", "nodePorts", conntrackCleanupServiceNodePorts.UnsortedList())
-	for _, nodePort := range conntrackCleanupServiceNodePorts.UnsortedList() {
-		filters = append(filters, filterForPort(nodePort, v1.ProtocolUDP))
+	entries, err := ct.ListEntries(ipFamilyMap[ipFamily])
+	if err != nil {
+		klog.ErrorS(err, "Failed to list conntrack entries")
+		return
 	}
 
-	if n, err := ct.ClearEntries(ipFamilyMap[ipFamily], filters...); err != nil {
-		klog.ErrorS(err, "Failed to delete stale service connections")
-	} else {
-		klog.V(4).InfoS("Deleted conntrack stale entries for services", "count", n)
+	// The indices "ipIndex" and "portIndex" represent allowed DNAT flows and are
+	// structured as maps for efficient lookups during cleanup. The key of the maps
+	// is origin (original destination [--orig-dst] for ipIndex and
+	// original port [--org-port-dst] for portIndex) and the value is set of allowed
+	// DNATed destinations (--reply-src).
+
+	// ipIndex maps service IPs (ClusterIP, LoadBalancerIPs and ExternalIPs) with the set of endpointIPs.
+	ipIndex := make(map[string]sets.Set[string])
+	// portIndex maps service NodePort with the set of endpointIPs.
+	portIndex := make(map[int]sets.Set[string])
+
+	for svcName, svc := range svcPortMap {
+		// we are only interested in UDP services
+		if svc.Protocol() != v1.ProtocolUDP {
+			continue
+		}
+
+		endpointIPs := sets.New[string]()
+		for _, endpoint := range endpointsMap[svcName] {
+			// Serving endpoints will never black-hole traffic, we consider all
+			// the serving endpoints as allowed DNATed destinations of service.
+			// Changes to the service should not affect existing flows, so we do
+			// not take traffic policies or topology of the service into account.
+			// This ensures that the behavior of UDP services remains consistent
+			// with TCP services.
+			if endpoint.IsServing() {
+				endpointIPs.Insert(endpoint.IP())
```

Comment:
Hm... I'm pretty sure this needs to take the port into account too, not just the IP. (Consider the case of changing the `targetPort` on a service; then endpointIP:oldPort would become invalid but endpointIP:newPort would be valid.)

(It looks like the existing code does this "wrong" too, so this could just be a FIXME. Or maybe we don't care; it's kind of an edge case.)

Reply:
good catch